aipmiddleware/validateparent.go

```diff
@@ -0,0 +1,89 @@
+package aipmiddleware
+
+import (
+	"context"
+
+	"go.einride.tech/aip/resourcename"
+	"google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// ParentValidator validates that response resources are children of requested parent resources.
+type ParentValidator struct {
+	// LogHook is called when parent validation fails, with the failing request/response, parent/name combination.
+	LogHook func(info *grpc.UnaryServerInfo, request, response proto.Message, parent, name string)
+}
+
+// UnaryServerInterceptor implements grpc.UnaryServerInterceptor.
+func (p *ParentValidator) UnaryServerInterceptor(
+	ctx context.Context,
+	request interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	protoRequest, ok := request.(proto.Message)
+	if !ok {
+		return nil, status.Error(codes.Internal, "parent validator got non-proto request")
+	}
+	protoReflectRequest := protoRequest.ProtoReflect()
+	// Request must have a parent field of string kind.
+	parentField := protoReflectRequest.Descriptor().Fields().ByName("parent")
+	if parentField == nil ||
+		parentField.IsMap() ||
+		parentField.IsList() ||
+		parentField.Kind() != protoreflect.StringKind {
+		return handler(ctx, request)
+	}
+	// The parent field must have a resource reference with a child type.
+	parentResourceReference := proto.GetExtension(
+		parentField.Options(), annotations.E_ResourceReference,
+	).(*annotations.ResourceReference)
+	if parentResourceReference.GetChildType() == "" {
+		return handler(ctx, request)
+	}
+	parent := protoReflectRequest.Get(parentField).String()
+	if parent == "" {
+		return handler(ctx, request)
+	}
+	response, err := handler(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	protoResponse, ok := response.(proto.Message)
+	if !ok {
+		return nil, status.Error(codes.Internal, "parent validator got non-proto response")
+	}
+	protoResponse.ProtoReflect().Range(func(field protoreflect.FieldDescriptor, value protoreflect.Value) bool {
+		if field.Kind() != protoreflect.MessageKind || !field.IsList() {
+			return true
+		}
+		fieldMessage := field.Message()
+		fieldMessageNameField := fieldMessage.Fields().ByName("name")
+		if fieldMessageNameField == nil || fieldMessageNameField.Kind() != protoreflect.StringKind {
+			return true
+		}
+		resourceDescriptor := proto.GetExtension(
+			fieldMessage.Options(), annotations.E_Resource,
+		).(*annotations.ResourceDescriptor)
+		if resourceDescriptor.GetType() != parentResourceReference.GetChildType() {
+			return true
+		}
```

Comment:
extract to `isResource(msg protoreflect.Message)`? This is pretty dense atm.

Reply:
ah, we use the child type as well. Maybe that signature needs some extra arguments. But I still think breaking some of this out is a good idea.

Reply:
Agree - I'll have a look at refactoring this method

Reply:
Done!